### PR TITLE
contracts: Update game state machine

### PR DESCRIPTION
**Pull request by j-mueller**
Change the endpoints to use the new contracts API, leave everything else as it was

### CODE REVIEW

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs [hunk 1]
```diff
@@ -146,92 +178,82 @@ gameTokenVal =
...
+lock :: AsContractError e => Contract GameStateMachineSchema e ()
+lock = do
+    LockArgs{lockArgsSecret, lockArgsValue} <- endpoint @"lock"
+    let secret            = HashedString (sha2_256 (C.pack lockArgsSecret))
+        addr              = Scripts.scriptAddress scriptInstance
+        initialDataScript = DataScript $ PlutusTx.toData (Initialised secret)
+    
+    -- 1. Create a transaction output with the value and the secret and
+    --    wait until it is confirmed.
+    _ <- both (writeTxSuccess $ Tx.payToScript lockArgsValue addr initialDataScript) (fundsAtAddressGt addr zero)
```
Comment (michaelpj) [inline plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs]: Why do we need this? Does `writeTxSuccess` not wait until the tx is confirmed?

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs [hunk 2]
```diff
@@ -146,92 +178,82 @@ gameTokenVal =
...
+                (Scripts.validatorScript scriptInstance)
+                redeemer
+
+             -- part 2: Lock the remaining funds in the contract with the new 
+             -- state 'ds'
+             Haskell.<> Tx.payToScript guessArgsValueLocked addr ds 
+
+             -- part 3: Pay the token to our own public key (this is a no-op on 
+             -- the wallet's balance but it ensures that the token is present 
+             -- in the transaction)
```
[the remark this answers is not in the document]
Comment (j-mueller) [reply plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs]: Yeah I completely agree, we need something like a "minimum value that's moved by this transaction", which could include the token.

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs (hunk 1 above)
Comment (j-mueller) [reply plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs]: the "success" in `writeTxSuccess` refers to the wallet successfully balancing and signing the transaction, and submitting it to the chain. It doesn't wait until the tx is confirmed. (I think we might need an `awaitTxConfirmed` effect for that)

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs [hunk 3]
```diff
@@ -146,92 +178,82 @@ gameTokenVal =
...
+        redeemer      = RedeemerScript $ PlutusTx.toData input
+        addr          = Scripts.scriptAddress scriptInstance
+
+    -- We need to know the unspent outputs at the contract address, and a public
+    -- key that we can send the token to. 
+    utxo <- utxoAt addr
+    pubKey <- ownPubKey
+
+    -- The transaction has three parts which are joined together using the 
+    -- 'Haskell.<>' semigroup operator.
```
[the remark this answers is not in the document]
Comment (j-mueller) [reply plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs]: Yup I'll do that

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs]: Yeah, and that should be more reliable than the awkward thing we do with tracking the amount at the address currently.

**Pull request merged**